ps2d/src/main.rs

```diff
@@ -79,114 +85,161 @@ impl<F: Fn(u8,bool) -> char> Ps2d<F> {
+                    if scroll & (1 << 3) == 1 << 3 {
+                        scroll -= 16;
                     }
+                    dz = -scroll as i32;
+                }
 
-                    if dz != 0 {
-                        self.input.write(&ScrollEvent {
-                            x: 0,
-                            y: dz,
-                        }.to_event()).expect("ps2d: failed to write scroll event");
-                    }
-                } else {
-                    println!("ps2d: overflow {:X} {:X} {:X} {:X}", self.packets[0], self.packets[1], self.packets[2], self.packets[3]);
+                let x = cmp::max(0, cmp::min(self.width as i32, self.mouse_x + dx));
+                let y = cmp::max(0, cmp::min(self.height as i32, self.mouse_y + dy));
+                if x != self.mouse_x || y != self.mouse_y {
+                    self.mouse_x = x;
+                    self.mouse_y = y;
+                    self.input.write(&MouseEvent {
+                        x: x,
+                        y: y,
+                    }.to_event()).expect("ps2d: failed to write mouse event");
+                }
+
+                let left = flags.contains(LEFT_BUTTON);
+                let middle = flags.contains(MIDDLE_BUTTON);
+                let right = flags.contains(RIGHT_BUTTON);
+                if left != self.mouse_left || middle != self.mouse_middle || right != self.mouse_right {
+                    self.mouse_left = left;
+                    self.mouse_middle = middle;
+                    self.mouse_right = right;
+                    self.input.write(&ButtonEvent {
+                        left: left,
+                        middle: middle,
+                        right: right,
+                    }.to_event()).expect("ps2d: failed to write button event");
                 }
 
-                self.packets = [0; 4];
-                self.packet_i = 0;
+                if dz != 0 {
+                    self.input.write(&ScrollEvent {
+                        x: 0,
+                        y: dz,
+                    }.to_event()).expect("ps2d: failed to write scroll event");
+                }
+            } else {
+                println!("ps2d: overflow {:X} {:X} {:X} {:X}", self.packets[0], self.packets[1], self.packets[2], self.packets[3]);
             }
+
+            self.packets = [0; 4];
+            self.packet_i = 0;
         }
     }
+
+    fn get_modifiers(&self) -> ModKeys {
+        let mut modifiers = ModKeys::empty();
+        if self.lshift { modifiers.insert(MOD_LSHIFT) }
+        if self.rshift { modifiers.insert(MOD_RSHIFT) }
+        if self.alt    { modifiers.insert(MOD_ALT) }
+        if self.alt_gr { modifiers.insert(MOD_ALT_GR) }
+        modifiers
+    }
+    /// Conversion from scancode to keycode when the scancode is only one byte
+    fn keycode_convert_single(scancode: u8) -> u8 {
+        scancode // At the moment, they are pretty much the same
```

Comment:
I think we should at least remap those keys that require escaping so all the codes fit into one byte.

If you do want to go down the linux keycode route though, feel free to use this bit of code from an experiment I did last week* that does the remapping: https://github.com/kaedroho/orbkb/blob/master/src/scancode.rs

\* I've been meaning to turn some of that work into a PR but ran out of time!